lib/teiserver_web/controllers/account/session_controller.ex

```diff
@@ -34,9 +34,62 @@ defmodule TeiserverWeb.Account.SessionController do
   def login(conn, %{"user" => %{"email" => email, "password" => password}}) do
     email = String.trim(email)
 
+    case UserLib.authenticate_user(conn, email, password) do
+      {:ok, user} ->
+        login_reply({:ok, user}, conn)
+
+      {:requires_2fa, user} ->
+        conn
+        |> put_session(:pending_2fa_user_id, user.id)
+        |> redirect(to: ~p"/otp")
+
+      {:error, reason} ->
+        login_reply({:error, reason}, conn)
+    end
+  end
+
+  def otp(conn, _params) do
+    user_id = get_session(conn, :pending_2fa_user_id)
+    user = Account.get_user!(user_id)
+
     conn
-    |> UserLib.authenticate_user(email, password)
-    |> login_reply(conn)
+    |> assign(:user, user)
+    |> render("totp.html")
+  end
+
+  def verify_totp(conn, %{"user_id" => user_id, "otp" => otp}) do
+    user = UserLib.get_user(user_id)
+
+    case Account.validate_totp(user, otp) do
+      {:ok, _} ->
+        delete_session(conn, :pending_2fa_user_id)
+        login_reply({:ok, user}, conn)
+
+      {:error, reason} ->
+        flash_message =
+          case reason do
+            :used ->
+              "OTP has already been used."
+
+            :invalid ->
+              "Invalid OTP."
```

Comment:
continuing on the previous comments, I think this is useless. A simple "Invalid code" is enough.
Also, this is going to be shown to user, the vast majority have zero clue what `OTP` means. For them, it's just a 6 digit code that refreshes often.